From 96322b5f707d47f953d3639ffd3bb1b44c11e40e Mon Sep 17 00:00:00 2001
From: Jiahao Xu <xujiahao@loongson.cn>
Date: Wed, 4 Jan 2023 11:35:45 +0800
Subject: [PATCH 24/28] Generate sw rsqrtf for better throughput under
 -ffinite-math-only and -funsafe-math-optimizations.

gcc/testsuite/ChangeLog:

        * gcc.target/loongarch/recip_sqrt.c: New test.

Change-Id: If328cac6fdeb2f2782afd9f8434e10953330ad5a
---
 src/gcc/config/loongarch/loongarch.md               |  2 +-
 src/gcc/testsuite/gcc.target/loongarch/recip_sqrt.c | 11 +++++++++++
 2 files changed, 12 insertions(+), 1 deletion(-)
 create mode 100644 src/gcc/testsuite/gcc.target/loongarch/recip_sqrt.c

diff --git a/src/gcc/config/loongarch/loongarch.md b/src/gcc/config/loongarch/loongarch.md
index 2c667107e..cddf76141 100644
--- a/src/gcc/config/loongarch/loongarch.md
+++ b/src/gcc/config/loongarch/loongarch.md
@@ -1025,7 +1025,7 @@
   [(set (match_operand:SF 0 "register_operand")
     (unspec:SF [(match_operand:SF 1 "register_operand")]
            UNSPEC_RSQRT))]
-  "flag_unsafe_math_optimizations && flag_rounding_math"
+  "flag_unsafe_math_optimizations && flag_finite_math_only"
 {
   loongarch_emit_swrsqrtsf (operands[0], operands[1], SFmode);
   DONE;
diff --git a/src/gcc/testsuite/gcc.target/loongarch/recip_sqrt.c b/src/gcc/testsuite/gcc.target/loongarch/recip_sqrt.c
new file mode 100644
index 000000000..ac1a8f177
--- /dev/null
+++ b/src/gcc/testsuite/gcc.target/loongarch/recip_sqrt.c
@@ -0,0 +1,11 @@
+/* { dg-do compile } */
+/* { dg-options "-Ofast -fdump-tree-optimized" } */
+
+float
+foo (float a)
+{
+  float tmp = 1.0f / __builtin_sqrtf (a);
+  return tmp;
+}
+
+/* { dg-final { scan-tree-dump-not " / " "optimized" } } */
-- 
2.31.1

